crates/exex/src/manager.rs

```diff
@@ -0,0 +1,446 @@
+        self.update_capacity();
+
+        // advance all poll senders
+        let mut min_id = usize::MAX;
+        for idx in (0..self.exex_handles.len()).rev() {
+            let mut exex = self.exex_handles.swap_remove(idx);
+
+            // it is a logic error for this to ever underflow since the manager manages the
+            // notification IDs
+            let notification_id = exex
+                .next_notification_id
+                .checked_sub(self.min_id)
+                .expect("exex expected notification ID outside the manager's range");
+            if let Some(notification) = self.buffer.get(notification_id) {
+                debug!(exex.id, notification_id, "sent notification to exex");
+                if let Poll::Ready(Err(err)) = exex.send(cx, notification) {
+                    // the channel was closed, which is irrecoverable for the manager
+                    return Poll::Ready(Err(err.into()))
+                }
+            }
+            min_id = min_id.min(exex.next_notification_id);
+            self.exex_handles.push(exex);
+        }
+
+        // remove processed buffered events
+        self.buffer.retain(|&(id, _)| id >= min_id);
+        self.min_id = min_id;
+        debug!("lowest notification id in buffer is {min_id}");
+
+        // update capacity
+        self.update_capacity();
+
+        // handle incoming exex events
+        for exex in self.exex_handles.iter_mut() {
+            while let Poll::Ready(Some(event)) = exex.receiver.poll_recv(cx) {
+                debug!(?event, "received event from exex {}", exex.id);
+                exex.metrics.events_sent_total.increment(1);
+                match event {
+                    ExExEvent::FinishedHeight(height) => exex.finished_height = Some(height),
+                }
+            }
+        }
+
+        // update watch channel block number
+        let finished_height = self.exex_handles.iter_mut().try_fold(u64::MAX, |curr, exex| {
+            let height = match exex.finished_height {
+                None => return Err(()),
+                Some(height) => height,
+            };
+
+            if height < curr {
+                Ok(height)
+            } else {
+                Ok(curr)
+            }
+        });
+        if let Ok(finished_height) = finished_height {
+            let _ = self.finished_height.send(Some(finished_height));
+        }
+
+        Poll::Pending
+    }
+}
+
+/// A handle to communicate with the [`ExExManager`].
+#[derive(Debug)]
+pub struct ExExManagerHandle {
+    exex_tx: UnboundedSender<CanonStateNotification>,
+    num_exexs: usize,
+    is_ready_receiver: watch::Receiver<bool>,
+    is_ready: WatchStream<bool>,
+    current_capacity: Arc<AtomicUsize>,
+    finished_height: watch::Receiver<Option<BlockNumber>>,
+}
+
+impl ExExManagerHandle {
+    /// Synchronously send a notification over the channel to all execution extensions.
+    ///
+    /// Senders should call [`Self::has_capacity`] first.
```

Comment:
should we call it inside this method instead, so that caller can't go beyond the capacity?